### PR TITLE
`xml.extension.jars` to add LemMinX extension jars

### LINKED ISSUE #251: Add LemMinx extensions via vscode user settings
LemMinx is extensible to manage custom completion, validation, etc. https://github.com/eclipse/lemminx-maven is a good sample of extension which provides support for maven.

To use this extension, it requires to create a vscode extension (eg: https://github.com/angelozerr/vscode-xml-maven)

This issue is to provide a new xml settings like `xml.extension.jars` which is an array of String which defines paths from LemMinx JAR extensions. to use when classpath is created

### PR DESCRIPTION
**Pull request by datho7561**
Adds a new setting `xml.extension.jars`, where you can list jars to contribute to the LemMinX classpath. You can use this feature to test LemMinX extensions in VSCode.

Closes #251

Signed-off-by: David Thompson <davthomp@redhat.com>

### CODE REVIEW

**Comment by datho7561** [comment]
I manually tested this with [the Liquidbase extension](https://github.com/Treehopper/liquibase-lsp)

**Comment by datho7561** [comment]
Need to update the internal docs with a better description

File: docs/Preferences.md [hunk 1]
```diff
@@ -161,3 +161,15 @@ The different options are:
...
+
+[LemMinX](https://github.com/eclipse/lemminx), the underlying language server that vscode-xml uses to provide most of its functionality,
+is extensible through Java's SPI mechanism.
+Please see [the extension development documentation](https://github.com/eclipse/lemminx/blob/master/docs/LemMinX-Extensions.md) for more information on how this works.
+
+JARs can be contributed to the LemMinX classpath using the `xml.extension.jars` preference.
+These paths can include globs.
+This feature is only intended to be used for LemMinX extension development purposes.
+Distributing vscode-xml extensions is best done through the mechanism described in the
+[vscode-xml extension development documentation](https://github.com/redhat-developer/vscode-xml/wiki/Extensions).
```
Comment (angelozerr) [inline docs/Preferences.md:175]: I think Extensions.md should be store  inside vscode-xml/docs/Extensions.md (embedded doc)

File: package.json [hunk 2]
```diff
@@ -355,7 +355,13 @@
...
           "type": "boolean",
           "default": true,
-          "markdownDescription": "Show referenced grammars in the Outline. Default is `true.",
+          "markdownDescription": "Show referenced grammars in the Outline. Default is `true`.",
+          "scope": "window"
+        },
+        "xml.extension.jars": {
+          "type": "array",
+          "default": [],
+          "markdownDescription": "An array of paths to JARs that should be contributed to the LemMinX classpath. This is intended to be used as a tool for developing extensions to vscode-xml. Please see [here](command:xml.open.docs?%5B%7B%22page%22%3A%22Preferences%22%2C%22section%22%3A%22extension-jars%22%7D%5D) for more information",
```
Comment (angelozerr) [inline package.json:364]: It seems that you can use glob pattern, please add that in the description.

File: docs/Preferences.md [hunk 3]
```diff
@@ -161,3 +161,15 @@ The different options are:
 Here is a demonstration of the effects of the setting on hovering. The above schema is used in the example:
 
 ![Changing the documentation type setting changes which text the hover shows when hovering over an element that is in a schema document](./images/Preferences/HoverDocumentationQuickDemo.gif)
+
+## Extension JARs
+
+[LemMinX](https://github.com/eclipse/lemminx), the underlying language server that vscode-xml uses to provide most of its functionality,
```
Comment (angelozerr) [inline docs/Preferences.md:167]: I think we should change with:

```
The LemMinX XML Language Server can be extended with custom plugins to provide additional validation and assistance. Typically this is done for specific files or contexts. External extensions are 
are contributed via an external JAR. For example [LemMinX Maven extension](https://github.com/eclipse/lemminx-maven/) provides extra assistance for the Maven pom.xml files.
Please see [the extension development documentation](https://github.com/eclipse/lemminx/blob/master/docs/LemMinX-Extensions.md) for more information on how this works.
...
```

File: docs/Preferences.md (hunk 1 above)
Comment (angelozerr) [inline docs/Preferences.md:175]: It should be nice if we could explain how to use this settings with lemminx-maven. I don't know if the lemminx-maven zip is available https://github.com/eclipse/lemminx-maven/pull/148#issuecomment-704780003

If it is available it should be nice to add a section here and provide a screenshot with completion maven.

Comment (angelozerr) [inline docs/Preferences.md:175]: I will move this Extensions.md in the docs (emebed documentation)

**Comment by angelozerr** [comment]
@datho7561 I think we should remove replace content of https://github.com/redhat-developer/vscode-xml#custom-xml-extensions with a link to the Extensions.md (not from the wiki but from the docs folder).

File: docs/Preferences.md (hunk 1 above)
Comment (datho7561) [reply docs/Preferences.md:175]: I don't know how to get lemminx-maven to work. Is there a way to generate an uber jar, or do I need to include all the separate jars with a glob?

Comment (datho7561) [reply docs/Preferences.md]: I tried loading all the jars in the generated zip from a project build, but it doesn't work. I get an exception about invalid certificates. Liquidbase works fine, though.

Comment (angelozerr) [reply docs/Preferences.md]: Could you report this issue in lemminx-maven please.

Comment (angelozerr) [reply docs/Preferences.md]: I would prefer explain how to configure lemminx-maven which is an Eclipse Project, but if it doesn't work for the moment, please create an issue for that (add lemminx-maven sample in the doc).

File: docs/Extensions.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
+# Custom XML Extensions
+
+The [LemMinX - XML Language Server](https://github.com/eclipse/lemminx) can be extended to support custom completion, hover, validation, rename, etc by using the [Java Service Provider Interface (SPI)](https://www.baeldung.com/java-spi) mechanism.
+vscode-xml provides the ability to use your custom XML support provider, by adding external jars to the XML language server's classpath.
+
+To do that:
+
+ * create a Java project which provides a custom XML extension providing your custom completion, hover, validation, rename, etc:
+ * create the XML extension like [MavenPlugin](https://github.com/eclipse/lemminx-maven/blob/master/lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxExtension.java).
```
Comment (angelozerr) [inline docs/Extensions.md:9]: [MavenLemminxExtension]

File: docs/Extensions.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+```json
+"contributes": {
+  "xml.javaExtensions": [
+    "./jar/dependencies/*.jar",
+    "./jar/my-xml-extension.jar"
+  ]
+}
+```
+
+You can see the [vscode-xml-maven](https://github.com/angelozerr/vscode-xml-maven) sample which registers custom maven completion [MavenCompletionParticipant](https://github.com/angelozerr/lsp4xml-extensions-maven/blob/master/org.eclipse.lsp4xml.extensions.maven/src/main/java/org/eclipse/lsp4xml/extensions/maven/MavenCompletionParticipant.java#L28) for scope:
```
Comment (angelozerr) [inline docs/Extensions.md:34]: Update the link of MavenCompletionParticipant to reference lemminx-maven.

**Review by angelozerr — APPROVED** [review]

**Comment by angelozerr** [comment]
Thanks @datho7561 !

**Pull request merged**